File: python/server/services/friend_service.py
from sqlalchemy import select, or_, and_
from sqlalchemy.ext.asyncio import AsyncSession

from ..models.user import User
from ..models.friend import Friend


async def send_friend_request(session: AsyncSession, user_id: int, friend_code: str):
    result = await session.execute(select(User).where(User.friend_code == friend_code))
    friend = result.scalar_one_or_none()
    if not friend or friend.id == user_id:
        return None
    existing = await session.execute(
        select(Friend).where(
            or_(
                and_(Friend.user_id == user_id, Friend.friend_id == friend.id),
                and_(Friend.user_id == friend.id, Friend.friend_id == user_id),
            )
        )
    )
    if existing.scalar_one_or_none():
        return None
    f1 = Friend(user_id=user_id, friend_id=friend.id, status="pending", direction="sent")
    f2 = Friend(user_id=friend.id, friend_id=user_id, status="pending", direction="received")
    session.add_all([f1, f2])
    await session.commit()
    return friend.id


async def respond_friend_request(session: AsyncSession, user_id: int, requester_id: int, accept: bool):
    result = await session.execute(
        select(Friend).where(
            Friend.user_id == user_id, Friend.friend_id == requester_id, Friend.status == "pending"
        )
    )
    entry = result.scalar_one_or_none()
    if not entry:
        return False
    entry.status = "accepted" if accept else "declined"
    other = await session.execute(
        select(Friend).where(
            Friend.user_id == requester_id, Friend.friend_id == user_id, Friend.status == "pending"
        )
    )
    other_entry = other.scalar_one_or_none()
    if other_entry:
        other_entry.status = entry.status
    await session.commit()
    return True


async def remove_friend(session: AsyncSession, user_id: int, friend_id: int):
    for col in [(user_id, friend_id), (friend_id, user_id)]:
        result = await session.execute(
            select(Friend).where(Friend.user_id == col[0], Friend.friend_id == col[1])
        )
        entry = result.scalar_one_or_none()
        if entry:
            await session.delete(entry)
    await session.commit()


async def get_friends(session: AsyncSession, user_id: int):
    result = await session.execute(
        select(Friend).where(Friend.user_id == user_id, Friend.status == "accepted")
    )
    return result.scalars().all()


async def get_friend_requests(session: AsyncSession, user_id: int):
    result = await session.execute(
        select(Friend).where(Friend.user_id == user_id, Friend.status == "pending", Friend.direction == "received")
    )
    return result.scalars().all()


async def get_outbound_requests(session: AsyncSession, user_id: int):
    result = await session.execute(
        select(Friend).where(Friend.user_id == user_id, Friend.status == "pending", Friend.direction == "sent")
    )
    return result.scalars().all()

File: js/apps/server/src/services/friendService.ts
import { prisma } from '../utils/database.js';
import { ApiError } from '../utils/http.js';
import { toFriendPayload, toPublicUser } from './serializers.js';
import { emitToUser } from '../socketio/emitter.js';
import type { Friend, User } from '@spotalong/types';

export async function sendFriendRequest(userId: string, friendCode: string): Promise<string> {
  const target = await prisma.user.findUnique({ where: { friendCode: friendCode.toUpperCase() } });
  if (!target) throw new ApiError(404, 'User not found');
  if (target.id === userId) throw new ApiError(400, 'You cannot add yourself as a friend');

  const existing = await prisma.friend.findFirst({
    where: {
      OR: [
        { userId, friendId: target.id },
        { userId: target.id, friendId: userId }
      ]
    }
  });
  if (existing) {
    throw new ApiError(
      400,
      existing.status === 'accepted' ? 'Already friends' : 'A request already exists'
    );
  }

  await prisma.friend.createMany({
    data: [
      { userId, friendId: target.id, direction: 'sent', status: 'pending' },
      { userId: target.id, friendId: userId, direction: 'received', status: 'pending' }
    ]
  });

  const requester = await prisma.user.findUnique({ where: { id: userId } });
  if (requester) {
    emitToUser(target.id, 'new_request', {
      userId: target.id,
      friendId: requester.id,
      direction: 'received',
      status: 'pending',
      otherUser: toPublicUser(requester)
    });
  }

  return target.id;
}

export async function respondToFriendRequest(
  userId: string,
  requesterId: string,
  action: 'accept' | 'decline'
): Promise<void> {
  const status = action === 'accept' ? 'accepted' : 'declined';
  const result = await prisma.friend.updateMany({
    where: {
      OR: [
        { userId, friendId: requesterId },
        { userId: requesterId, friendId: userId }
      ],
      status: 'pending'
    },
    data: { status }
  });

  if (result.count === 0) throw new ApiError(404, 'Request not found');

  if (action === 'accept') {
    const [me, other] = await Promise.all([
      prisma.user.findUnique({ where: { id: userId } }),
      prisma.user.findUnique({ where: { id: requesterId } })
    ]);
    if (me && other) {
      emitToUser(userId, 'new_friend', toPublicUser(other));
      emitToUser(requesterId, 'new_friend', toPublicUser(me));
    }
  } else {
    emitToUser(requesterId, 'remove_request', { requesterId });
  }
}

export async function removeFriend(userId: string, friendCode: string): Promise<void> {
  const target = await prisma.user.findUnique({ where: { friendCode: friendCode.toUpperCase() } });
  if (!target) throw new ApiError(404, 'User not found');

  await prisma.listenSession.deleteMany({
    where: {
      OR: [
        { listenerId: userId, targetId: target.id },
        { listenerId: target.id, targetId: userId }
      ]
    }
  });

  await prisma.friend.deleteMany({
    where: {
      OR: [
        { userId, friendId: target.id },
        { userId: target.id, friendId: userId }
      ]
    }
  });

  const [me] = await Promise.all([prisma.user.findUnique({ where: { id: userId } })]);
  if (me) {
    emitToUser(target.id, 'remove_friend', toPublicUser(me));
  }
}

export interface FriendsData {
  friends: User[];
  friendRequests: Friend[];
  outboundFriendRequests: Friend[];
}

export async function getFriendsData(userId: string): Promise<FriendsData> {
  const rows = await prisma.friend.findMany({
    where: {
      OR: [{ userId }, { friendId: userId }]
    },
    include: { user: true, friend: true }
  });

  const friends: User[] = [];
  const friendRequests: Friend[] = [];
  const outboundFriendRequests: Friend[] = [];

  for (const row of rows) {
    if (row.userId !== userId) continue;

    if (row.status === 'accepted') {
      friends.push(toPublicUser(row.friend));
    } else if (row.status === 'pending') {
      if (row.direction === 'sent') {
        outboundFriendRequests.push(toFriendPayload(row));
      } else {
        friendRequests.push(toFriendPayload(row));
      }
    }
  }

  return { friends, friendRequests, outboundFriendRequests };
}
